src/st_sbardef.c

```diff
@@ -0,0 +1,511 @@
+        maxwidth = MAX(maxwidth, SHORT(out->characters[i]->width));
+        maxheight = MAX(maxheight, SHORT(out->characters[i]->height));
+    }
+
+    out->maxheight = maxheight;
+
+    switch (out->type)
+    {
+        case sbf_mono0:
+            out->monowidth = SHORT(out->characters[0]->width);
+            break;
+        case sbf_monomax:
+            out->monowidth = maxwidth;
+            break;
+        default:
+            break;
+    }
+
+    return true;
+}
+
+static boolean ParseStatusBar(json_t *json, statusbar_t *out)
+{
+    json_t *height = JS_GetObject(json, "height");
+    json_t *fullscreenrender = JS_GetObject(json, "fullscreenrender");
+    if (!JS_IsNumber(height) || !JS_IsBoolean(fullscreenrender))
+    {
+        return false;
+    }
+    out->height = JS_GetInteger(height);
+    out->fullscreenrender = JS_GetBoolean(fullscreenrender);
+
+    out->fillflat = JS_GetStringCopy(json, "fillflat");
+
+    json_t *js_children = JS_GetObject(json, "children");
+    json_t *js_child = NULL;
+    JS_ArrayForEach(js_child, js_children)
+    {
+        sbarelem_t elem = {0};
+        if (ParseSbarElem(js_child, &elem))
+        {
+            array_push(out->children, elem);
+        }
+    }
+
+    return true;
+}
+
+sbardef_t *ST_ParseSbarDef(void)
+{
+    json_t *json = JS_Open("SBARDEF", "statusbar", (version_t){1, 1, 0});
+    if (json == NULL)
+    {
+        return NULL;
+    }
+
+    boolean load_defaults = false;
+    version_t v = {0};
+    JS_GetVersion(json, &v);
+    if (v.major == 1 && v.minor < 1)
+    {
+        load_defaults = true;
+    }
+
+    json_t *data = JS_GetObject(json, "data");
+    if (JS_IsNull(data) || !JS_IsObject(data))
+    {
+        I_Printf(VB_ERROR, "SBARDEF: no data");
+        JS_Close(json);
+        return NULL;
+    }
+
+    sbardef_t *out = calloc(1, sizeof(*out));
+
+    json_t *js_numberfonts = JS_GetObject(data, "numberfonts");
+    json_t *js_numberfont = NULL;
+
+    JS_ArrayForEach(js_numberfont, js_numberfonts)
+    {
+        numberfont_t numberfont = {0};
+        if (ParseNumberFont(js_numberfont, &numberfont))
+        {
+            array_push(out->numberfonts, numberfont);
+        }
+    }
+
+    json_t *js_hudfonts = JS_GetObject(data, "hudfonts");
+    json_t *js_hudfont = NULL;
+
+    JS_ArrayForEach(js_hudfont, js_hudfonts)
+    {
+        hudfont_t hudfont = {0};
+        if (ParseHUDFont(js_hudfont, &hudfont))
+        {
+            array_push(out->hudfonts, hudfont);
+        }
+    }
+
+    json_t *js_statusbars = JS_GetObject(data, "statusbars");
+    json_t *js_statusbar = NULL;
+
+    JS_ArrayForEach(js_statusbar, js_statusbars)
+    {
+        statusbar_t statusbar = {0};
+        if (ParseStatusBar(js_statusbar, &statusbar))
+        {
+            array_push(out->statusbars, statusbar);
+        }
+    }
+
+    JS_Close(json);
+
+    if (!load_defaults)
+    {
+        return out;
+    }
+
+    json = JS_Open("SBHUDDEF", "hud", (version_t){1, 0, 0});
```

Comment:
If we allow to load SBARDEF lumps with version < 1.1.0 (i.e. without HUD widget definitions) and extend them by loading our own HUD widget definitions here, then why do we have to extend our own SBARDEF lump with HUD widget definitions and increase its version to 1.1.0 in the first place?

Reply:
Yes, we can move all widgets to our own lump. But it's more convenient to have them integrated in SBARDEF, for example the user can have different `canvas` elements with different conditions, containing different STBAR and HUD widgets.

We also support Boom translation lumps and additional widescreen alignment, so we should increase the version anyway.